src/spotify-service.ts

```diff
@@ -1,67 +1,78 @@
 import { SpotifyApi, type AccessToken } from "@spotify/web-api-ts-sdk";
-import { Context, Effect, Layer, Secret, Encoding, pipe, Queue } from "effect";
-import { SpotifyConfigService } from "./spotify-config-service";
+import { Context, Data, Effect, Encoding, Layer, Queue, Secret } from "effect";
 import { Message, MessagePubSub } from "./message-pubsub";
+import { SpotifyConfigService } from "./spotify-config-service";
+
+export class SpotifyError extends Data.TaggedError("SpotifyError")<{
+  cause: unknown;
+}> {}
+
+const make = Effect.gen(function* () {
+  const config = yield* SpotifyConfigService;
+
+  const client = SpotifyApi.withAccessToken(
+    config.clientId,
+    config.accessToken
+  );
+  const use = <A>(f: (client: SpotifyApi) => Promise<A>) =>
+    Effect.tryPromise({
+      try: () => f(client),
+      catch: (cause) => new SpotifyError({ cause }),
+    });
+
+  return { use, client } as const;
+});
 
 export class SpotifyApiClient extends Context.Tag("spotify-api-client")<
   SpotifyApiClient,
-  SpotifyApi
+  Effect.Effect.Success<typeof make>
 >() {
-  static Live = Layer.scoped(
-    this,
-    Effect.gen(function* () {
-      const config = yield* SpotifyConfigService;
-
-      return SpotifyApi.withAccessToken(config.clientId, config.accessToken);
-    }),
-  ).pipe(Layer.provide(SpotifyConfigService.Live));
+  static Live = Layer.scoped(this, make).pipe(
+    Layer.provide(SpotifyConfigService.Live)
+  );
 }
 
-export const SpotifyService = Layer.effectDiscard(
-  Effect.scoped(
-    Effect.gen(function* () {
-      yield* Effect.logInfo("starting spotify service");
-      const api = yield* SpotifyApiClient;
-      const pubsub = yield* MessagePubSub;
-
-      const dequeue = yield* pubsub.subscribeTo("CurrentlyPlayingRequest");
+export const SpotifyService = Layer.scopedDiscard(
```

Comment:
You can use Layer.scopedDiscard instead of Effect.scoped here

Reply:
TIL